running-code/execute-package.md

```diff
@@ -0,0 +1,163 @@
+and will execute it as a script.
+
+This execution mode is performed with the `-m` flag, as in `python -m site`. It can be used in place of a file
+path, but cannot be used in combination with a path, as there can only be one executing module.
+
+:::{tip}
+These commands both do the same thing, but the latter is easier to remember and much more portable, as it doesn't rely
+on the local machine's file system details.
+
+```bash
+python ./.venv/lib/python3.12/site-packages/pip/__main__.py
+```
+
+```bash
+python -m pip
+```
+:::
+
+### Further exploration
+
+On your own or in small groups:
+
+* Install the `my_program.py` module from the [last lesson](execute-script-launch-command)
+* Try to get the same greeting as before using `python -m my_program`.
+
+**my_program.py** which prints out `✨ Hello from Python ✨` when executed as a script.
+
+```python
+#!/usr/bin/env python
+# The above line is a shebang and can take the place of typing python on the command line
+# This comment is below, because shebangs must be the first line!
+
+def shiny_hello():
+    print("\N{Sparkles} Hello from Python \N{Sparkles}")
+
+if __name__ == "__main__":
+    shiny_hello()
+```
+
+### Executable packages
+
+The `-m` flag as described above only works for Python modules (files), but does not work for Python (sub-)packages (directories).
+This means that you cannot execute a command using only the name of your package when it is structured to use directories
+
+Once your package grows, the top-level name `my_program` turns into a directory.
+(See [Python Package Structure](https://www.pyopensci.org/python-package-guide/package-structure-code/python-package-structure.html)
+for when and how to create a package structure).
+```
+project/
+└── src/
+    └── my_program/
+        ├── __init__.py
+        └── greeting.py
+```
+
+However, you can never execute a directory.
+```bash
+python -m my_program
+python: No module named my_program.__main__; 'my_program' is a package and cannot be directly executed
+```
+
+Initially, Python seems to tell you that the directory names, including your top-level package name,
+cannot be directly executed. But the error message contains the hint that you need to make this run properly.
+
+[Earlier you learned](execute-script-name-eq-main) that `if __name__ == "__main__":` can protect parts of your
+script from executing when it is imported, making that conditional change the file's behavior when used as a script.
+There is a very similar concept that can be applied to Python packages.
+
+Any package that contains a [`__main__.py` module](https://docs.python.org/3/library/__main__.html#module-__main__)
+can be executed directly using the `python` command, without reference to any specific module files.
+
+:::{note}
+The `__main__.py` file typically doesn't have an `if __name__ == "__main__":` conditional in it, as its execution
+is already separated out from the rest of the package.
```

Comment:
```suggestion
The `__main__.py` file typically doesn't have an `if __name__ == "__main__":` conditional in it, as its execution
is already separated out from the rest of the package.
The `__init__.py` file will be run whenever the package is imported,
but the `__main__.py` file will only be run [when the package is run from the command line](https://docs.python.org/3/library/__main__.html#main-py-in-python-packages) with the `-m` flag.
```

keeping parallelism with previous section - "why do we have `if __name__ == "__main__"`? to keep code from running when imported" part 2. also being explicit about when code is executed as scaffolding

Reply:
maybe, trying not to open the door to an import machinery tutorial with `__init__.py`. I could probably write ten lessons on that and still be wrong on some of it.

Reply:
that's fair - i was thinking from a "continuity with thinking about packages as modules" perspective not like "try and be logically complete" pov, but i see what you mean and if ya think it's tmi then np to close this <3